engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,575 @@ package ingestion
-func (s *StopControl) GetState() StopControlState {
-	s.RLock()
-	defer s.RUnlock()
-	return s.state
+	log := sc.log.With().
+		Bool("node_will_react_to_version_beacon",
+			sc.nodeVersion != nil).
+		Logger()
+
+	if sc.nodeVersion != nil {
+		log = log.With().
+			Stringer("node_version", sc.nodeVersion).
+			Bool("crash_on_version_boundary_reached",
+				sc.crashOnVersionBoundaryReached).
+			Logger()
+	}
+
+	log.Info().Msgf("Created")
+
+	// TODO: handle version beacon already indicating a stop
+	// right now the stop will happen on first BlockFinalized
+	// which is fine, but ideally we would stop right away
+
+	return sc
 }
 
-// IsPaused returns true is block execution has been paused
-func (s *StopControl) IsPaused() bool {
+// IsExecutionStopped returns true is block execution has been stopped
+func (s *StopControl) IsExecutionStopped() bool {
 	s.RLock()
 	defer s.RUnlock()
-	return s.state == StopControlPaused
+
+	return s.stopped
 }
 
-// SetStopHeight sets new stopHeight and crash mode, and return old values:
-//   - stopHeight
-//   - crash
-//
-// Returns error if the stopping process has already commenced, new values will be rejected.
-func (s *StopControl) SetStopHeight(
-	height uint64,
-	crash bool,
-) (uint64, bool, error) {
+// SetStop sets new stop parameters.
+// Returns error if the stopping process has already commenced, or if already stopped.
+func (s *StopControl) SetStop(
+	stop StopParameters,
+) error {
 	s.Lock()
 	defer s.Unlock()
 
-	oldHeight := s.stopHeight
-	oldCrash := s.crash
-
-	if s.state == StopControlCommenced {
-		return oldHeight,
-			oldCrash,
-			fmt.Errorf(
-				"cannot update stopHeight, "+
-					"stopping commenced for stopHeight %d with crash=%t",
-				oldHeight,
-				oldCrash,
-			)
+	stopBoundary := &stopBoundary{
+		StopParameters: stop,
 	}
 
-	if s.state == StopControlPaused {
-		return oldHeight,
-			oldCrash,
-			fmt.Errorf("cannot update stopHeight, already paused")
+	log := s.log.With().
+		Stringer("old_stop", s.stopBoundary).
+		Stringer("new_stop", stopBoundary).
+		Logger()
+
+	canChange, reason := s.canChangeStop(
+		stopBoundary.StopHeight,
+		false,
+	)
+	if !canChange {
+		err := fmt.Errorf(reason)
+
+		log.Warn().Err(err).Msg("cannot set stopHeight")
+		return err
 	}
 
-	// cannot set stopHeight to block which is already executing
-	// so the lowest possible stopHeight is highestExecutingHeight+1
-	if height <= s.highestExecutingHeight {
-		return oldHeight,
-			oldCrash,
-			fmt.Errorf(
-				"cannot update stopHeight, "+
-					"given stopHeight %d below or equal to highest executing height %d",
-				height,
-				s.highestExecutingHeight,
-			)
+	log.Info().Msg("stop set")
+	s.stopBoundary = stopBoundary
+
+	return nil
+}
+
+// canChangeStop verifies if the stop parameters can be changed
+// returns false and the reason if the parameters cannot be changed
+// if newHeight == math.MaxUint64 tha basically means that the stop is being removed
```

Comment:
if EN has executed block 100, and is previously set to stop at 200, and now we are changing it to 50, it should ignore, right?

Reply:
I don't think so. I think it should just stop immediately. That is the way its written now.

My argument is that if you were trying to manually stop 2 nodes, and while you sent the command to stop the first one in 200 blocks, the second one could already be more than 200 blocks ahead, In which case I would prefer that is still stops right away instead of failing to set a stop at all.